lang/cs/Org.Apache.REEF.Network/Group/Task/Impl/GroupCommNetworkObserver.cs

```diff
@@ -16,84 +16,81 @@
 // under the License.
 
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Linq;
 using Org.Apache.REEF.Network.Group.Driver.Impl;
 using Org.Apache.REEF.Network.NetworkService;
 using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Implementations.InjectionPlan;
+using Org.Apache.REEF.Utilities.Diagnostics;
 using Org.Apache.REEF.Utilities.Logging;
+using Org.Apache.REEF.Wake.Remote;
+using Org.Apache.REEF.Wake.Remote.Impl;
 
 namespace Org.Apache.REEF.Network.Group.Task.Impl
 {
     /// <summary>
     /// Handles all incoming messages for this Task.
     /// Writable version
     /// </summary>
-    internal sealed class GroupCommNetworkObserver : IGroupCommNetworkObserver
+    internal sealed class GroupCommNetworkObserver : IObserver<IRemoteMessage<NsMessage<GeneralGroupCommunicationMessage>>>
     {
-        private static readonly Logger LOGGER = Logger.GetLogger(typeof(GroupCommNetworkObserver));
+        private static readonly Logger Logger = Logger.GetLogger(typeof(GroupCommNetworkObserver));
 
-        private readonly Dictionary<string, IObserver<GeneralGroupCommunicationMessage>> _commGroupHandlers;
+        private readonly IInjectionFuture<StreamingNetworkService<GeneralGroupCommunicationMessage>> _networkService;
+
+        private readonly ConcurrentDictionary<string, TaskMessageObserver> _taskMessageObservers =
+            new ConcurrentDictionary<string, TaskMessageObserver>();
+
+        private readonly ConcurrentDictionary<string, byte> _registeredNodes = new ConcurrentDictionary<string, byte>();
 
         /// <summary>
         /// Creates a new GroupCommNetworkObserver.
         /// </summary>
         [Inject]
-        private GroupCommNetworkObserver()
+        private GroupCommNetworkObserver(
+            IInjectionFuture<StreamingNetworkService<GeneralGroupCommunicationMessage>> networkService)
```

Comment:
Can you clarify the semantics of IInjectionFuture? I looked at REEF code but it has no documentation. What is the benefit of using it vs. directly the injection. Looking at the implementation, one thing I see is the thread safety for instantiation.